src/agentscope/service/web_search/web_digest.py

```diff
@@ -0,0 +1,298 @@
+            `html_parse_func` will be invoked with the response as
+            input. For example,
+                `html_parse_func(requests.get(url))`
+
+    Returns:
+        `ServiceResponse`: If successful, `ServiceResponse` object is returned
+        with `content` field is a dict, where keys are `html_parsing_types` and
+        values are the (pre-processed) parsed content. For example, if
+        `html_parsing_types = ['raw', 'selected_tags_to_text']` the
+        `content` field is
+        {
+            "raw": raw_content_of_web,
+            "selected_tags_to_text":
+                processed_text_content_of_the_selected_tags,
+        }
+    """
+    logger.info(f"parsing_types: {html_parsing_types}")
+    assert set(html_parsing_types) <= set(
+        typing.get_args(HTML_PARSING_TYPES),
+    ), (
+        f"html_parsing_types must be subset of "
+        f"{typing.get_args(HTML_PARSING_TYPES)}"
+    )
+
+    results = {}
+
+    for parse_type in html_parsing_types:
+        if parse_type == "raw":
+            results[str(parse_type)] = html_response.text
+        elif parse_type == "self_define_func" and html_parse_func is not None:
+            results[str(parse_type)] = html_parse_func(html_response)
+        elif parse_type == "selected_tags_to_text":
+            logger.info(
+                f"extracting text information "
+                f"from tags: {html_selected_tags}",
+            )
+            try:
+                from bs4 import BeautifulSoup, NavigableString, Tag
+            except ImportError as exc:
+                raise ImportError(
+                    "BeautifulSoup4 is required for processing the "
+                    "web page without model."
+                    "Please install with `pip install bs4` .",
+                ) from exc
+
+            doc = BeautifulSoup(html_response.text, "html.parser")
+
+            def get_navigable_strings(
+                e: Tag,
+            ) -> str:
+                # pylint: disable=cell-var-from-loop
+                text = ""
+                for child in e.children:
+                    if isinstance(child, Tag):
+                        # pylint: disable=cell-var-from-loop
+                        text += get_navigable_strings(child).strip(" \n\t")
+                    elif isinstance(child, NavigableString):
+                        if (e.name == "a") and (href := e.get("href")):
+                            if is_valid_url(href):
+                                text += f"{child.strip()} ({href})"
+                        else:
+                            text += child.text
+                return " ".join(text.split())
+
+            text_parts = ""
+            if html_selected_tags is None:
+                html_selected_tags = ["h", "p", "li", "div", "a"]
+            for element in doc.find_all(recursive=True):
+                if element.name in html_selected_tags:
+                    text_parts += get_navigable_strings(element).strip(" \n\t")
```

Comment:
If the tags here are nested, will it cause the same content to be extracted multiple times?